docs-guides/source/opt-stable-diffusion.md

```diff
@@ -0,0 +1,188 @@
+
+```python
+import coremltools as ct
+import coremltools.optimize as cto
+
+mlmodel = ct.models.MLModel(
+    "output-xl-512/Stable_Diffusion_version_stabilityai_stable-diffusion-xl-base-1.0_unet.mlpackage"
+)
+op_config = cto.coreml.OpPalettizerConfig(
+            nbits=4,
+            mode="kmeans",
+            granularity="per_grouped_channel",
+            group_size=16, 
+)
+
+config = cto.coreml.OptimizationConfig(    
+    global_config=op_config,
+)
+compressed_mlmodel = cto.coreml.palettize_weights(mlmodel, config)
+```
+
+Experimenting with group_size equal to 4, 16 & 32 we see the following results:
+
+| Config                | Model Size | Image                                                               |
+|-----------------------|------------|---------------------------------------------------------------------|
+| 4-bit (per-tensor)    | 1.21 GB    | <img height="250px" src="images/sdxl-4bit.png" width="250px"/>      | 
+| 4-bit (group_size=4)  | 1.22 GB    | <img height="250px" src="images/sdxl-4bit-gs4.png" width="250px"/>  |
+| 4-bit (group_size=16) | 1.21 GB    | <img height="250px" src="images/sdxl-4bit-gs16.png" width="250px"/> |
+| 4-bit (group_size=32) | 1.21 GB    | <img height="250px" src="images/sdxl-4bit-gs32.png" width="250px"/> |
+
+We see that images generated with grouped channel palettization are much better than one generated with scalar palettization. We are able to reclaim most of the accuracy back, with a very slight increase in size. This increase is due to the little bit of extra space taken by the additional lookup tables.
+                                                                                 
+## Joint compression 
+
+In this section, we aim to further compress the model using joint compression techniques. We'll utilize the `ct.optimize.torch.*` APIs to compress the PyTorch model. 
+The process involves two main steps: pruning the model using calibration data with [SparseGPT](opt-pruning-algos.md#sparsegpt), followed by data-free palettization of the pruned model using [PostTrainingPalettizer](opt-palettization-algos.md#k-means).
+
+### Pruning with SparseGPT
+
+The `LayerwiseCompressor` (`SparseGPT`) algorithm compresses a sequential model layer by layer. Since it only supports sequential models, where the output of one layer directly feeds into the input of the next, some minor modifications are required for the `UNet2DConditionModelXL` to apply this algorithm.
+
+We adjust the forward interface for the down, mid, and up blocks to ensure that the `hidden_states`, `down_block_res_samples`, `temb`, and `encoder_hidden_states` flow through all the blocks. The updated `hidden_states` and `down_block_res_samples` are then returned as outputs. This adjustment is necessary to avoid any skip connections across these blocks.
```

Comment:
Why do we want to avoid skip connections?

Reply:
I tried to explain that in the previous para: 
> The `LayerwiseCompressor` (`SparseGPT`) algorithm compresses a sequential model layer by layer. Since it only supports sequential models, where the output of one layer directly feeds into the input of the next, some minor modifications are required for the `UNet2DConditionModelXL` to apply this algorithm.

Reply:
oh, ok that makes sense now. Maye just change the last sentence to say something like: "This adjustment is necessary to avoid any skip connections across these blocks, so the  model is sequential."

Reply:
Done